src/vocutouts/uws/butler.py

```diff
@@ -26,13 +31,51 @@ class UWSButler:
         The Butler repository to use.
     """
 
-    def __init__(self, repository: str) -> None:
+    def __init__(self, repository: str, config: UWSConfig) -> None:
         self._butler = Butler(repository)
+        self._config = config
+        self._gcs = storage.Client()
+
+    async def url_for_result(self, result: JobResult) -> JobResultURL:
+        """Query Butler for the URL for a job result.
+
+        Notes
+        -----
+        Currently, this is synchronous, blocking the rest of the cutout
+        frontend, and probably painfully slow (although I've not gotten hard
+        measurements yet).  Butler has to be refreshed to pick up new
+        collections (and, the first time we do a cutout, the new data type),
+        which does synchronous database calls.  Butler is not designed to be
+        used concurrently (so far as I know) so running these calls in a
+        thread pool is probably unsafe.
 
-    def url_for_result(self, result: JobResult) -> str:
-        """Query Butler for the URL for a job result."""
-        return self._butler.getURI(
+        This is a stop-gap until we have client/server Butler, so perhaps we
+        can live with it until then.
+        """
+        self._butler.registry.refresh()
+        uri = self._butler.getURI(
             result.datatype,
             dataId=result.data_id,
             collections=[result.collection],
-        ).geturl()
+        )
+
+        # Generate a signed URL for the result.  Eventually, we may need to
+        # cache these so that we don't regenerate them on every retrieval of
+        # the results.
+        assert uri.scheme == "s3"
+        bucket = self._gcs.bucket(uri.netloc)
+        blob = bucket.blob(uri.path[1:])
```

Comment:
In theory `uri.relativeToPathRoot()` could be used here but your way is obviously shorter.

Reply:
Ah, thanks, I missed that.